### PR TITLE
Move shell aware env into spack.environment.shell

### PR DESCRIPTION
**Pull request by haampie**
Cherry-picked from #25564 so this is standalone.

With this PR we can activate an environment in Spack itself, without computing changes to environment variables only necessary for "shell aware" env activation.

1. Activating an environment:
    
    ```python
    spack.environment.activate(Environment(xyz)) -> None
    ```
    this basically just sets `_active_environment` and modifies some config scopes.

2. Activating an environment **and** getting environment variable modifications for the shell:

    ```python
    spack.environment.shell.activate(Environment(xyz)) -> EnvironmentModifications
    ```

This should make it easier/faster to do unit tests and scripting with spack, without the cli interface.

### CODE REVIEW

**Comment by alalazo** [comment]
Can you rebase and solve the conflict? I'll review immediately after that!

File: lib/spack/spack/environment/__init__.py
Comment (alalazo) [inline lib/spack/spack/environment/__init__.py]: I wonder if methods that are related to environment views should better be extracted into a separate module like `spack.environment.views`, reducing the number of responsibilities of the `Environment` class. Not necessarily a request for this PR, but I'd like to start the discussion since we are looking for a refactor.

Comment (alalazo) [reply lib/spack/spack/environment/__init__.py]: In any case I think taking an `EnvironmentModification` as an argument is an improvement of the API imo.

File: lib/spack/spack/environment/__init__.py
Comment (alalazo) [inline lib/spack/spack/environment/__init__.py]: What would be the issue with the simpler:
```python
env = active_environment()
try:
    deactivate()
    yield
finally:
    if env:
        activate(env)
```
since `deactivate` should handle the case of no active environment?

File: lib/spack/spack/environment/shell.py
Comment (alalazo) [inline lib/spack/spack/environment/shell.py]:
```suggestion
    Deactivate an environment and collect corresponding environment modifications
```

File: lib/spack/spack/util/environment.py
Comment (alalazo) [inline lib/spack/spack/util/environment.py]: Why do we have this return value? This method operates solely on side effects - it modifies the `env` passed as an argument or `os.environ`. Where is it useful to chain calls on this method? 

I think returning a value gives the wrong impression that the object passed as argument and the return value are independent and this can be combined in a functional style.

File: lib/spack/spack/test/cmd/env.py
Comment (alalazo) [inline lib/spack/spack/test/cmd/env.py]: This is what I was referring to. Reading these two lines one could think `apply_modifications` computes a new return value, while instead it's modifying the literal `{}` passed as argument. I think we should:
```suggestion
    env_variables = {}
    spack.environment.shell.activate(e).apply_modifications(env_variables)
    assert env_variables['DEPENDENCY_ENV_VAR'] == '1'
```
to be more explicit on what is happening underneath.

**Review by alalazo — COMMENTED** [review]
I gave a first pass at this refactor. I like moving shell commands to a sub-package and  I only left some minor request at the moment.

I have a preference to rename `spack.environment.__init__.py` to `spack.environment.environment.py` and leave `__init__.py` with just a few lines of code that glue together the most commonly used functionalities from sub-packages similarly to:
```python
from .environment import Environment
...

__all__ = [
    'Environment',
    ...
]
```
The idea would be to have as little logic as possible in `__init__.py` and use relative imports for sub-packages. Might be worth checking with other core devs if we share a common opinion on this before going full steam ahead.

**Review by alalazo — CHANGES_REQUESTED** [review]
Sorry, should have been request for changes :laughing:

File: lib/spack/spack/util/environment.py
Comment (haampie) [reply lib/spack/spack/util/environment.py]: hm yeah, makes sense

Comment (haampie) [reply lib/spack/spack/util/environment.py]: I don't really like the fact that this defaults to `os.environ` though, can't we just make it a required argument and drop the `env = os.environ`?

Comment (haampie) [reply lib/spack/spack/util/environment.py]: I'm just gonna undo the changes here

File: lib/spack/spack/environment/__init__.py
Comment (haampie) [reply lib/spack/spack/environment/__init__.py]: Yeah, ties in with #25622

File: lib/spack/spack/environment/environment.py
Comment (alalazo) [inline lib/spack/spack/environment/environment.py]: Minor, but I don't like the -ed name here. What about:
```suggestion
def no_active_environment():
```
if the previous name seemed too "imperative"?

File: lib/spack/spack/environment/environment.py
Comment (alalazo) [inline lib/spack/spack/environment/environment.py:2078]: Would it be hard to fix this? Does it require anything more than keeping track of this option somewhere and tie it to a local variable in the context manager?

File: lib/spack/spack/environment/environment.py
Comment (alalazo) [inline lib/spack/spack/environment/environment.py]: I would put this after the global declaration. It's also worth updating documentation to point out the semantics changed and a `None` argument results in a no-op.

File: lib/spack/spack/main.py
Comment (alalazo) [inline lib/spack/spack/main.py]: Not sure what to do here. The fact that `activate` now accepts a `None` is good, since it makes the method symmetric with deactivate. On the caller side though it was more readable before, since now a reader needs to know that:
1. `env` can be `None`
2. A `None` argument to `activate` is a no-op

Can we update the comment above and make to document this behavior at the calling site?

File: lib/spack/spack/subprocess_context.py
Comment (alalazo) [inline lib/spack/spack/subprocess_context.py]: :+1:

**Review by alalazo — CHANGES_REQUESTED** [review]
Left a few minor comments. I like the simplification in unit-tests.

File: lib/spack/spack/main.py
Comment (haampie) [reply lib/spack/spack/main.py]: In that case it's maybe best to just undo the activate-accepting-none-changes

File: lib/spack/spack/environment/environment.py
Comment (haampie) [reply lib/spack/spack/environment/environment.py:2078]: Hm, it can be stored in the env itself. I was also thinking that it wouldn't hurt ot get rid of the flag, and instead put this in the spack.yaml, which is more transparent than remembering whether or not you had this --use-env-repo set enabled or not.

Comment (haampie) [reply lib/spack/spack/environment/environment.py:2078]: but that's for another pr...

Comment (alalazo) [reply lib/spack/spack/environment/environment.py:2078]: Fair enough.

File: lib/spack/spack/main.py
Comment (alalazo) [reply lib/spack/spack/main.py]: If we go back on this, can we remove it from `deactivate` too? It seems really odd that you can:
```python
deactivate(None)
```
but not:
```python
activate(None)
```
What do you think?

Comment (alalazo) [reply lib/spack/spack/main.py]: Or we can add an additional Boolean argument to activate/deactivate to accept or not `None`.

Comment (haampie) [reply lib/spack/spack/main.py]: Done that

File: lib/spack/spack/environment/environment.py
Comment (haampie) [reply lib/spack/spack/environment/environment.py]: fixed

File: lib/spack/spack/environment/__init__.py
Comment (alalazo) [inline lib/spack/spack/environment/__init__.py]: Can we sort this one the same as isort did above?

File: lib/spack/spack/test/env.py
Comment (alalazo) [inline lib/spack/spack/test/env.py]: Typo:
```suggestion
def test_activate_should_require_an_env():
```
?

File: lib/spack/spack/cmd/spec.py
Comment (alalazo) [inline lib/spack/spack/cmd/spec.py]: I would revert this change and add an epilog:
```python
subparser.epilog = "When an environment is active and no specs are provided, the environment root specs are used instead."
```

**Review by alalazo — COMMENTED** [review]
Only minor requests, but basically LGTM. Pinging other reviewers.

File: lib/spack/spack/environment/__init__.py
Comment (haampie) [reply lib/spack/spack/environment/__init__.py]: Oh, I thought isort handled this too with --fix, let me change it

File: lib/spack/spack/test/env.py
Comment (haampie) [reply lib/spack/spack/test/env.py]: ah yes, thanks

File: lib/spack/spack/cmd/spec.py
Comment (haampie) [reply lib/spack/spack/cmd/spec.py]: Thanks, didn't know that existed

Comment (haampie) [reply lib/spack/spack/cmd/spec.py]: Huh, this was a commit pushed to the wrong branch :sweat_smile:

Comment (haampie) [reply lib/spack/spack/cmd/spec.py]: Reverted, I'll take this as your review of #25941 then :p

**Comment by haampie** [comment]
@spackbot run pipeline

File: lib/spack/spack/environment/environment.py
Comment (alalazo) [inline lib/spack/spack/environment/environment.py:1238]: Docstring are missing

File: lib/spack/spack/environment/environment.py
Comment (alalazo) [inline lib/spack/spack/environment/environment.py:1266]: Docstring are missing

**Review by alalazo — COMMENTED** [review]
Noticed two missing docstrings, but otherwise LGTM.

File: lib/spack/spack/environment/environment.py
Comment (haampie) [reply lib/spack/spack/environment/environment.py:1238]: used to be missing too (add_default_view_to_shell), but sure, i can add them

Comment (haampie) [reply lib/spack/spack/environment/environment.py:1238]: honestly I don't really understand what this function even does :laughing:

File: lib/spack/spack/environment/environment.py
Comment (alalazo) [inline lib/spack/spack/environment/environment.py:1249]: Minor, so can be done in a following PR. This method modifies the state of `env_mod` so it should not return it (gives the false impression that the return value is a different object than the argument). It's minor since peeking at code it seems we always discard the return value.

File: lib/spack/spack/environment/environment.py
Comment (alalazo) [inline lib/spack/spack/environment/environment.py:1277]: Same as above.

**Review by alalazo — APPROVED** [review]

**Review by tgamblin — APPROVED** [review]

File: lib/spack/spack/environment/environment.py
Comment (haampie) [reply lib/spack/spack/environment/environment.py:1249]: it's a bit like a builder pattern

Comment (alalazo) [reply lib/spack/spack/environment/environment.py:1249]: It's not, I think. Here you pass the object as an argument and you return the same object. In the builder patter you have a method to get the result and others to build parts, but none of them takes as argument the object to be modified and returns it modified. 

With this:
```python
env1 = ...
env2 = ev.add_default_view_to_env(env1)
```
`env1` and `env2` are the same object. You modify one and you see the other modified too.

Comment (haampie) [reply lib/spack/spack/environment/environment.py:1249]: i agree, it makes more sense not to hint it is side effect free

**Pull request merged**
